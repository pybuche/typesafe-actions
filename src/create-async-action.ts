import {
  TypeConstant,
  ActionCreatorBuilder,
  // ActionBuilderMap,
} from './type-helpers';
import { checkInvalidActionTypeInArray } from './utils/validation';
import { createAction } from './create-action';

export type AsyncActionCreator<
  TRequest extends [T1, P1],
  TSuccess extends [T2, P2],
  TFailure extends [T3, P3],
  TCancel extends [T4, P4] = never,
  T1 extends TypeConstant = TRequest[0],
  P1 = TRequest[1],
  T2 extends TypeConstant = TSuccess[0],
  P2 = TSuccess[1],
  T3 extends TypeConstant = TFailure[0],
  P3 = TFailure[1],
  T4 extends TypeConstant = TCancel[0],
  P4 = TCancel[1]
> = {
  request: ActionCreatorBuilder<T1, P1>;
  success: ActionCreatorBuilder<T2, P2>;
  failure: ActionCreatorBuilder<T3, P3>;
  cancel: TCancel extends [TypeConstant, any]
    ? ActionCreatorBuilder<T4, P4>
    : never;
};

export interface AsyncActionBuilder<
  TType1 extends TypeConstant,
  TType2 extends TypeConstant,
  TType3 extends TypeConstant,
  TType4 extends TypeConstant
> {
  <TPayload1, TPayload2, TPayload3, TPayload4>(): AsyncActionCreator<
    [TType1, TPayload1],
    [TType2, TPayload2],
    [TType3, TPayload3],
    [TType4, TPayload4]
  >;
  <TPayload1, TPayload2, TPayload3>(): AsyncActionCreator<
    [TType1, TPayload1],
    [TType2, TPayload2],
    [TType3, TPayload3]
  >;
}

/**
 * implementation
 */
export function createAsyncAction<
  TType1 extends TypeConstant,
  TType2 extends TypeConstant,
  TType3 extends TypeConstant,
  TType4 extends TypeConstant
>(
  requestType: TType1,
  successType: TType2,
  failureType: TType3,
  cancelType?: TType4
): AsyncActionBuilder<TType1, TType2, TType3, TType4> {
  [requestType, successType, failureType].forEach(
    checkInvalidActionTypeInArray
  );

  const constructor = (<TPayload1, TPayload2, TPayload3, TPayload4>() => {
    return {
      request: createAction(requestType)<TPayload1>(),
      success: createAction(successType)<TPayload2>(),
      failure: createAction(failureType)<TPayload3>(),
      cancel: cancelType && createAction(cancelType)<TPayload4>(),
    };
  }) as AsyncActionBuilder<TType1, TType2, TType3, TType4>;

  const api = Object.assign<
    AsyncActionBuilder<TType1, TType2, TType3, TType4>,
    {}
  >(constructor, {
    // extension point for chain api
  });

  return api;
}

// export interface AsyncActionBuilder<
// {
// withMappers<A1 = undefined, P1 = undefined, A2 = undefined, P2 = undefined, A3 = undefined, P3 = undefined>(
//   requestMapper: (a?: A1) => P1,
//   successMapper: (a?: A2) => P2,
//   failureMapper: (a?: A3) => P3
// ): AsyncActionBuilderWithMappers<T1, T2, T3, A1, P1, A2, P2, A3, P3>;
// }

// export type AsyncActionBuilderWithMappers<
//   T1 extends StringType,
//   T2 extends StringType,
//   T3 extends StringType,
//   A1 = undefined,
//   P1 = undefined,
//   A2 = undefined,
//   P2 = undefined,
//   A3 = undefined,
//   P3 = undefined
//   > = {
//     request: ActionBuilderMap<T1, A1, P1>;
//     success: ActionBuilderMap<T2, A2, P2>;
//     failure: ActionBuilderMap<T3, A3, P3>;
//   };

// function withMappers<A1, P1, A2, P2, A3, P3>(
//   requestMapper: (a?: A1) => P1,
//   successMapper: (a?: A2) => P2,
//   failureMapper: (a?: A3) => P3
// ): AsyncActionBuilderWithMappers<T1, T2, T3, A1, P1, A2, P2, A3, P3> {
//   return {
//     request: createCustomAction(requestType, type => (payload?: A1) => ({
//       type,
//       payload: requestMapper != null ? requestMapper(payload) : undefined,
//     })) as MapBuilder<T1, A1, P1>,
//     success: createCustomAction(successType, type => (payload?: A2) => ({
//       type,
//       payload: successMapper != null ? successMapper(payload) : undefined,
//     })) as MapBuilder<T2, A2, P2>,
//     failure: createCustomAction(failureType, type => (payload?: A3) => ({
//       type,
//       payload: failureMapper != null ? failureMapper(payload) : undefined,
//     })) as MapBuilder<T3, A3, P3>,
//   };
// }
