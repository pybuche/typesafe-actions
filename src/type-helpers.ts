/**
 * PUBLIC API
 */

/**
 * @desc Interface for internal types augmentation
 * @example ```
 * declare module 'typesafe-actions' {
 *   export type RootAction = ActionType<typeof import('./root-action').default>;
 *   export interface Types {
 *     RootAction: RootAction;
 *   }
 * } ```
 */
export interface Types {}

/**
 * @desc Type representing Type Constant
 */
export type TypeConstant = string;

/**
 * @desc Type representing Generic Action
 */
export type Action<TType extends TypeConstant = TypeConstant> = {
  type: TType;
};

/**
 * @desc Type representing Generic ActionCreator
 */
export type ActionCreator<TType extends TypeConstant = TypeConstant> = (
  ...args: any[]
) => Action<TType>;

/**
 * @desc Type representing Generic Reducer
 */
export type Reducer<TState, TAction extends Action> = (
  state: TState | undefined,
  action: TAction
) => TState;

/**
 * @desc Action without Payload
 */
export type EmptyAction<TType extends TypeConstant> = {
  type: TType;
};

/**
 * @desc Action with only Payload
 */
export type PayloadAction<TType extends TypeConstant, TPayload> = {
  type: TType;
  payload: TPayload;
};

// /**
//  * @desc Action with only Meta
//  */
// export type MetaAction<TType extends TypeConstant, TMeta> = {
//   type: TType;
//   meta: TMeta;
// };

/**
 * @desc Action with both Payload and Meta
 */
export type PayloadMetaAction<TType extends TypeConstant, TPayload, TMeta> = {
  type: TType;
  payload: TPayload;
  meta: TMeta;
};

/**
 * @desc Action Creator producing EmptyAction
 */
export type EmptyActionCreator<TType extends TypeConstant> = () => EmptyAction<
  TType
>;

/**
 * @desc Action Creator producing PayloadAction
 */
export type PayloadActionCreator<TType extends TypeConstant, TPayload> = (
  payload: TPayload
) => PayloadAction<TType, TPayload>;

// /**
//  * @desc Action Creator producing MetaAction
//  */
// export type MetaActionCreator<TType extends TypeConstant, TMeta> = (
//   payload: undefined,
//   meta: TMeta
// ) => MetaAction<TType, TMeta>;

/**
 * @desc Action Creator producing PayloadMetaAction
 */
export type PayloadMetaActionCreator<
  TType extends TypeConstant,
  TPayload,
  TMeta
> = (
  payload: TPayload,
  meta: TMeta
) => PayloadMetaAction<TType, TPayload, TMeta>;

/**
 * @desc Type representing type getter on Action Creator instance
 */
export interface ActionCreatorTypeMetadata<TType extends TypeConstant> {
  getType?: () => TType;
}

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  TActionCreatorOrMap extends any
> = TActionCreatorOrMap extends ActionCreator<TypeConstant>
  ? ReturnType<TActionCreatorOrMap>
  : TActionCreatorOrMap extends Record<any, any>
  ? {
      [K in keyof TActionCreatorOrMap]: ActionType<TActionCreatorOrMap[K]>;
    }[keyof TActionCreatorOrMap]
  : TActionCreatorOrMap extends infer R // TODO: should be just never but compiler yell with circularly references itself error
  ? never
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<
  TReducerOrMap extends any
> = TReducerOrMap extends Reducer<any, any>
  ? ReturnType<TReducerOrMap>
  : TReducerOrMap extends Record<any, any>
  ? { [K in keyof TReducerOrMap]: StateType<TReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/** @private */
export type ActionCreatorBuilder<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadActionCreator<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaActionCreator<TType, TPayload, TMeta>
      : EmptyActionCreator<TType>
    : PayloadActionCreator<TType, TPayload>
  : PayloadMetaActionCreator<TType, TPayload, TMeta>;

/** @private */
export type ActionBuilder<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadAction<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaAction<TType, TPayload, TMeta>
      : EmptyAction<TType>
    : PayloadAction<TType, TPayload>
  : PayloadMetaAction<TType, TPayload, TMeta>;

/** @private */
export type ResolveType<T extends any> = T extends Function
  ? T
  : { [K in keyof T]: T[K] };
