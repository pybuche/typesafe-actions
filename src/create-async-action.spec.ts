import * as TH from './type-helpers';
import { createAsyncAction, AsyncActionCreator } from './create-async-action';

type User = { firstName: string; lastName: string };

// @dts-jest:group async action with undefined type
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE'
  )<undefined, User[], Error>();

  const fn = (
    a: AsyncActionCreator<
      ['FETCH_USERS_REQUEST', undefined],
      ['FETCH_USERS_SUCCESS', User[]],
      ['FETCH_USERS_FAILURE', Error]
    >
  ) => a;
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);

  // @dts-jest:pass:snap
  fetchUsersAsync.request(); /* => {
    type: 'FETCH_USERS_REQUEST'
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [{ firstName: 'Piotr', lastName: 'Witek' }]
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    Error('reason')
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: Error('reason')
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel;
}

// @dts-jest:group async action with any type
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE'
  )<any, any[], any>();

  // @dts-jest:pass:snap
  fetchUsersAsync.request(
    1
  ); /* => {
    type: 'FETCH_USERS_REQUEST', payload: 1,
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    1,
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [1],
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    1
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: 1,
  } */
}

// @dts-jest:group async action with cancel
{
  const fetchUsersAsync = createAsyncAction(
    'FETCH_USERS_REQUEST',
    'FETCH_USERS_SUCCESS',
    'FETCH_USERS_FAILURE',
    'FETCH_USERS_CANCEL'
  )<undefined, User[], Error, string>();

  const fn = (
    a: AsyncActionCreator<
      ['FETCH_USERS_REQUEST', undefined],
      ['FETCH_USERS_SUCCESS', User[]],
      ['FETCH_USERS_FAILURE', Error],
      ['FETCH_USERS_CANCEL', string]
    >
  ) => a;
  // @dts-jest:pass:snap
  fn(fetchUsersAsync);

  // @dts-jest:pass:snap
  fetchUsersAsync.request(); /* => {
    type: 'FETCH_USERS_REQUEST'
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.success([
    { firstName: 'Piotr', lastName: 'Witek' },
  ]); /* => {
    type: 'FETCH_USERS_SUCCESS', payload: [{ firstName: 'Piotr', lastName: 'Witek' }]
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.failure(
    Error('reason')
  ); /* => {
    type: 'FETCH_USERS_FAILURE', payload: Error('reason')
  } */

  // @dts-jest:pass:snap
  fetchUsersAsync.cancel(
    'reason'
  ); /* => {
    type: 'FETCH_USERS_CANCEL', payload: 'reason'
  } */
}

// @dts-jest:group async action with mappers
// {
//   const fetchUserMappers = createAsyncAction(
//     'FETCH_USER_REQUEST',
//     'FETCH_USER_SUCCESS',
//     'FETCH_USER_FAILURE'
//   ).withMappers(
//     (id: string) => ({ id }), // request mapper
//     ({ firstName, lastName }: User) => `${firstName} ${lastName}`, // success mapper
//     () => 'hardcoded error' // error mapper
//   );

//   const requestResult: {
//     type: 'FETCH_USER_REQUEST';
//     payload: { id: string };
//   } = fetchUserMappers.request('fake_id');
//   expect(requestResult).toEqual({ type: 'FETCH_USER_REQUEST', payload: { id: 'fake_id' } });
//   const successResult: {
//     type: 'FETCH_USER_SUCCESS';
//     payload: string;
//   } = fetchUserMappers.success({
//     firstName: 'Piotr',
//     lastName: 'Witek',
//   });
//   expect(successResult).toEqual({
//     type: 'FETCH_USER_SUCCESS',
//     payload: 'Piotr Witek',
//   });
//   const failureResult: {
//     type: 'FETCH_USER_FAILURE';
//     payload: string;
//   } = fetchUserMappers.failure();
//   expect(failureResult).toEqual({
//     type: 'FETCH_USER_FAILURE',
//     payload: 'hardcoded error',
//   });
// }

// @dts-jest:group async action with unions
// {
//   const fetchUserMappers = createAsyncAction(
//     'FETCH_USER_REQUEST',
//     'FETCH_USER_SUCCESS',
//     'FETCH_USER_FAILURE'
//   ).withMappers(
//     (id: string | number) => ({ id }), // request mapper
//     ({ firstName, lastName }: User | undefined) => `${firstName} ${lastName}`, // success mapper
//     (error: boolean | string) => error // error mapper
//   );

//   const requestResult: {
//     type: 'FETCH_USER_REQUEST';
//     payload: { id: string | number };
//   } = fetchUserMappers.request(2);
//   expect(requestResult).toEqual({ type: 'FETCH_USER_REQUEST', payload: { id: 2 } });
//   const successResult: {
//     type: 'FETCH_USER_SUCCESS';
//     payload: string | undefined;
//   } = fetchUserMappers.success(undefined);
//   expect(successResult).toEqual({
//     type: 'FETCH_USER_SUCCESS',
//     payload: undefined,
//   });
//   const failureResult: {
//     type: 'FETCH_USER_FAILURE';
//     payload: boolean | string;
//   } = fetchUserMappers.failure(true);
//   expect(failureResult).toEqual({
//     type: 'FETCH_USER_FAILURE',
//     payload: true,
//   });
// }

// NEW API PROPOSAL:
// - WITH MAP
// const getTodoAsyncMap = createAsyncAction(
//   'GET_TODO_REQUEST',
//   'GET_TODO_SUCCESS',
//   'GET_TODO_FAILURE'
// ).map(...);
