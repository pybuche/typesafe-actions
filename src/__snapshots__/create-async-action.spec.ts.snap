// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async action with any type fetchUsersAsync.failure(
      1
    ) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_FAILURE\\", any>"`;

exports[`async action with any type fetchUsersAsync.request(
      1
    ) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_REQUEST\\", any>"`;

exports[`async action with any type fetchUsersAsync.success([
      1,
    ]) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_SUCCESS\\", any[]>"`;

exports[`async action with cancel fetchUsersAsync.cancel(
      'reason'
    ) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_CANCEL\\", string>"`;

exports[`async action with cancel fetchUsersAsync.failure(
      Error('reason')
    ) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_FAILURE\\", Error>"`;

exports[`async action with cancel fetchUsersAsync.request() 1`] = `"TH.EmptyAction<\\"FETCH_USERS_REQUEST\\">"`;

exports[`async action with cancel fetchUsersAsync.success([
      { firstName: 'Piotr', lastName: 'Witek' },
    ]) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_SUCCESS\\", User[]>"`;

exports[`async action with cancel fn(fetchUsersAsync) 1`] = `"AsyncActionCreator<[\\"FETCH_USERS_REQUEST\\", undefined], [\\"FETCH_USERS_SUCCESS\\", User[]], [\\"FETCH_USERS_FAILURE\\", Error], [\\"FETCH_USERS_CANCEL\\", string], \\"FETCH_USERS_REQUEST\\", undefined, \\"FETCH_USERS_SUCCESS\\", User[], \\"FETCH_USERS_FAILURE\\", Error, \\"FETCH_USERS_CANCEL\\", string>"`;

exports[`async action with undefined type fetchUsersAsync.cancel 1`] = `"never"`;

exports[`async action with undefined type fetchUsersAsync.failure(
      Error('reason')
    ) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_FAILURE\\", Error>"`;

exports[`async action with undefined type fetchUsersAsync.request() 1`] = `"TH.EmptyAction<\\"FETCH_USERS_REQUEST\\">"`;

exports[`async action with undefined type fetchUsersAsync.success([
      { firstName: 'Piotr', lastName: 'Witek' },
    ]) 1`] = `"TH.PayloadAction<\\"FETCH_USERS_SUCCESS\\", User[]>"`;

exports[`async action with undefined type fn(fetchUsersAsync) 1`] = `"AsyncActionCreator<[\\"FETCH_USERS_REQUEST\\", undefined], [\\"FETCH_USERS_SUCCESS\\", User[]], [\\"FETCH_USERS_FAILURE\\", Error], never, \\"FETCH_USERS_REQUEST\\", undefined, \\"FETCH_USERS_SUCCESS\\", User[], \\"FETCH_USERS_FAILURE\\", Error, never, never>"`;
